### PR TITLE
Synchronize accesses to reachability handlers

### PR DESCRIPTION
**Pull request by zakkak**
Prevents data races in reachability handlers registration when using
`-H:-RunReachabilityHandlersConcurrently`.

The fact that https://github.com/oracle/graal/issues/5868 is not deterministic and it's related to enabling/disabling concurrency hints that it's probably caused by a data race. 
Adding a synchonized blocks in `com.oracle.svm.hosted.ReachabilityHandlerFeature#registerReachabilityHandler(org.graalvm.nativeimage.hosted.Feature.BeforeAnalysisAccess, java.lang.Object, java.lang.Object[])` seems to resolve the issue.

Closes https://github.com/oracle/graal/issues/5868

### CODE REVIEW

**Review by christianwimmer — CHANGES_REQUESTED** [review]
Instead of the explicit `synchronized` blocks, it would be better to use proper synchronized data structures. For `ReachabilityHandlerFeature.activeHandlers` and `triggeredHandlers` that would be wrapping using `Collections.synchronizedMap`, while for the nested set that does not require identity semantics we can just use `ConcurrentHashMap.createKeySet`.

**Comment by zakkak** [comment]
Hi @christianwimmer I see there is `com.oracle.svm.core.util.ConcurrentIdentityHashMap` as well, shall I use it instead of `Collections.synchronizedMap` for better performance?

**Comment by christianwimmer** [comment]
> Hi @christianwimmer I see there is `com.oracle.svm.core.util.ConcurrentIdentityHashMap` as well, shall I use it instead of `Collections.synchronizedMap` for better performance?

Yes, I forgot that we have that.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java [hunk 1]
```diff
@@ -119,7 +122,7 @@ public void duringAnalysis(DuringAnalysisAccess a) {
                 Set<Object> triggers = activeHandlers.get(callback);
                 if (callback instanceof Consumer) {
                     if (isTriggered(access, triggers)) {
-                        triggeredHandlers.put(callback, null);
+                        triggeredHandlers.put(callback, NULL_MAP);
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java:125]: Why not just use `Map.of()`? We want that map to be immutable anyways.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java [hunk 2]
```diff
@@ -52,8 +54,9 @@
 @AutomaticallyRegisteredFeature
 public class ReachabilityHandlerFeature extends ReachabilityHandler implements InternalFeature {
 
-    private final IdentityHashMap<Object, Set<Object>> activeHandlers = new IdentityHashMap<>();
-    private final IdentityHashMap<Object, Map<Object, Set<Object>>> triggeredHandlers = new IdentityHashMap<>();
+    private static final Map<Object, Set<Object>> NULL_MAP = new IdentityHashMap<>();
+    private final ConcurrentIdentityHashMap<Object, ConcurrentHashMap.KeySetView<Object, Boolean>> activeHandlers = new ConcurrentIdentityHashMap<>();
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java:58]: Why change the generic type from `Set` to `ConcurrentHashMap.KeySetView`? The field can just be defined as `Map<Object, Set<Object>> activeHandlers`.

Same for all the other type definitions.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java [hunk 3]
```diff
@@ -88,7 +91,7 @@ private void registerReachabilityHandler(BeforeAnalysisAccess a, Object callback
         BeforeAnalysisAccessImpl access = (BeforeAnalysisAccessImpl) a;
         AnalysisMetaAccess metaAccess = access.getMetaAccess();
 
-        Set<Object> triggerSet = activeHandlers.computeIfAbsent(callback, c -> new HashSet<>());
+        ConcurrentHashMap.KeySetView<Object, Boolean> triggerSet = activeHandlers.computeIfAbsent(callback, c -> ConcurrentHashMap.newKeySet());
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java:94]: that's actually the prototypical use case for `var triggerSet`, which we can now use in the native-image code base.

**Review by christianwimmer — CHANGES_REQUESTED** [review]

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java (hunk 1 above)
Comment (zakkak) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java]: I didn't think about it. It makes much more sense. Done, thanks.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java (hunk 2 above)
Comment (zakkak) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java]: I guess I just like explicit types :)

Changed.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java (hunk 3 above)
Comment (zakkak) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ReachabilityHandlerFeature.java]: Done.

**Comment by christianwimmer** [comment]
@oubidar-Abderrahim can you please create the internal PR for this

**Review by christianwimmer — APPROVED** [review]

**Comment by zakkak** [comment]
Thank you for the review @christianwimmer

**Comment by zakkak** [comment]
This should also be backported to 23.1 and 23.0

**Comment by zakkak** [comment]
@oubidar-Abderrahim is there any progress in regards to getting this merged? Thanks

**Comment by christianwimmer** [comment]
> @oubidar-Abderrahim is there any progress in regards to getting this merged? Thanks

I'm in the process of running the final tests and benchmarks, should be merged soon. And it will be backported.

**Comment by zakkak** [comment]
Great, thanks for the update @christianwimmer.

**Comment by btomala** [comment]
We encountered the same issue. Thanks for fixing it. @christianwimmer when do we expect this will be released?

**Comment by zakkak** [comment]
@btomala it should land in [GraalVM for JDK 21.0.2 in late January](https://www.graalvm.org/release-calendar/).

**Pull request merged**